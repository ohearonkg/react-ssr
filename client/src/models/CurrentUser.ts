/**
 * The current user information
 * recieved from our api endpoint
 */
export interface ICurrentUser {
  _id: string;
  googleId: string;
  __v: number;
}
