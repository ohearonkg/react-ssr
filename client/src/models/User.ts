/**
 * The user object as obtained
 * from our endpoint
 */
export interface IUser {
  id: number;
  name: string;
}
