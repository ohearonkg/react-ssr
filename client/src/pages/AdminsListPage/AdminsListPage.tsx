import { AxiosInstance } from "axios";
import * as React from "react";
import { Helmet } from "react-helmet";
import { connect } from "react-redux";
import { AnyAction, Dispatch, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { fetchAdmins, fetchAdminsFunction } from "../../actions/Admins/admins";
import { fetchUsersFunction } from "../../actions/Users/users";
import RequireAuth from "../../hocs/RequireAuth/RequireAuth";
import { IUser } from "../../models/User";
import { ApplicationState } from "../../reducers";

interface IAdminsListPageProps {
  /**
   * Function to be called to fetch
   * the list of admins
   */
  fetchAdmins: () => void;

  /**
   * List of admins
   */
  admins: IUser[];
}

class AdminsListPage extends React.Component<IAdminsListPageProps, {}> {
  public componentDidMount() {
    this.props.fetchAdmins();
  }

  public renderAdmins = () =>
    this.props.admins.map((admin: IUser) => (
      <li key={admin.id}> {admin.name} </li>
    ));

  public buildHead = () => {
    return (
      <Helmet>
        <title> React SRR - Admins </title>
        <meta name="og:title" content="React SSR - Home" />
      </Helmet>
    );
  };

  public render() {
    return (
      <div>
        {this.buildHead()}
        <h1> Protected Admins List</h1>
        <ul>{this.renderAdmins()}</ul>
      </div>
    );
  }
}

/* Thunk called to load data on the
 * server side
 *
 * Arguments passed in include the axios instance
 * which is different between client and server. The
 * client side requests MUST be proxied.
 */
const loadData = (
  store: Store<ApplicationState, AnyAction> & {
    dispatch: ThunkDispatch<{}, AxiosInstance, AnyAction>;
  }
) => store.dispatch(fetchAdminsFunction());

const mapStateToProps = (state: ApplicationState) => {
  return {
    admins: state.admins.admins
  };
};

export { AdminsListPage };

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    // @ts-ignore
    fetchAdmins: () => dispatch(fetchAdminsFunction())
  };
};

export default {
  component: connect(
    mapStateToProps,
    mapDispatchToProps
  )(RequireAuth(AdminsListPage)),
  loadData
};
