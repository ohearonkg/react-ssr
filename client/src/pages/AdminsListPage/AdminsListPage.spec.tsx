import { shallow, ShallowWrapper } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { IUser } from "../../models/User";
import { AdminsListPage } from "./AdminsListPage";

describe("Admins List Page", () => {
  const sampleAdmins: IUser[] = [
    {
      id: 1,
      name: "admin1"
    },
    {
      id: 2,
      name: "admin2"
    }
  ];

  let wrapper: ShallowWrapper;
  beforeEach(() => {
    wrapper = shallow(
      <AdminsListPage
        fetchAdmins={() => {
          return;
        }}
        admins={[]}
      />
    );
  });

  /**
   * Snapshot
   */
  it("Should match its snapshot", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  /**
   * Rendering list of admins
   */
  it("Should render each user in its admins prop", () => {
    wrapper.setProps({ admins: sampleAdmins });

    expect(wrapper.find("li")).toHaveLength(2);

    const wrapperHtml: string = wrapper.html();
    expect(wrapperHtml).toMatch("admin1");
    expect(wrapperHtml).toMatch("admin2");
  });
});
