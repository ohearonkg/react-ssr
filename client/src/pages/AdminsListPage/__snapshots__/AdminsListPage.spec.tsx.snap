// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admins List Page Should match its snapshot 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
       React SRR - Admins 
    </title>
    <meta
      content="React SSR - Home"
      name="og:title"
    />
  </HelmetWrapper>
  <h1>
     Protected Admins List
  </h1>
  <ul />
</div>
`;
